File: rust/src/leetcode/Q213_house_robber_ii.rs
/*
 * @lc app=leetcode id=213 lang=rust
 *
 * [213] House Robber II
 */

// @lc code=start
impl Solution {
    fn rob_linear(nums: &[i32]) -> i32 {
        if nums.len() == 1 {
            return nums[0];
        }

        let mut profit: Vec<i32> = vec![0; nums.len()];
        profit[0] = nums[0];
        profit[1] = nums[0].max(nums[1]);

        for i in 2..nums.len() {
            profit[i] = (profit[i - 1]).max(profit[i - 2] + nums[i]);
        }

        profit[profit.len() - 1]
    }

    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.len() == 1 {
            return nums[0];
        }

        Self::rob_linear(&nums[0..nums.len() - 1]).max(Self::rob_linear(&nums[1..]))
    }
}
// @lc code=end

File: python/Q213_house_robber_ii.py
#
# @lc app=leetcode id=213 lang=python3
#
# [213] House Robber II
#


# @lc code=start
class Solution:
    def rob(self, nums: list[int]) -> int:
        if len(nums) <= 2:
            return max(nums)

        dp1: list[int] = [0] * (len(nums) - 1)
        dp1[0] = nums[0]
        dp1[1] = max(nums[0], nums[1])

        dp2: list[int] = [0] * (len(nums) - 1)
        dp2[0] = nums[1]
        dp2[1] = max(nums[1], nums[2])

        for i in range(2, len(nums) - 1):
            rob: int = nums[i] + dp1[i - 2]
            not_rob: int = dp1[i - 1]
            dp1[i] = rob if rob > not_rob else not_rob

            rob: int = nums[i + 1] + dp2[i - 2]
            not_rob: int = dp2[i - 1]
            dp2[i] = rob if rob > not_rob else not_rob

        return max(dp1[-1], dp2[-1])


# @lc code=end
